File: ios/LEDLamp/LampState.swift
//
//  LampState.swift
//  LEDLamp
//
//  Created by Oleksandr Fedko on 08.12.2024.
//

import Foundation

struct LampState {
    enum Mode: UInt8, CaseIterable {
        case color, rainbow
        
        var name: String {
            switch self {
            case .color: "💡 Color"
            case .rainbow: "🌈 Rainbow"
            }
        }
    }
    
    enum Error: LocalizedError {
        case invalidMode
    }
    
    var red: Double
    var green: Double
    var blue: Double
    var mode: Mode
    
    var data: Data {
        var buffer = Data()
        buffer.append(UInt8(red))
        buffer.append(UInt8(green))
        buffer.append(UInt8(blue))
        buffer.append(mode.rawValue)
        return buffer
    }
    
    static func deserialize(_ data: Data) throws -> LampState {
        var offset = 0
        let red = data[offset]
        offset += 1
        let green = data[offset]
        offset += 1
        let blue = data[offset]
        offset += 1
        guard let mode = LampState.Mode(rawValue: data[offset]) else { throw LampState.Error.invalidMode }
        return LampState(red: Double(red), green: Double(green), blue: Double(blue), mode: mode)
    }
}

File: android/app/src/main/java/com/example/ledlamp/LampState.kt
package com.example.ledlamp

data class LampState(
    val red: Float = 0f,
    val green: Float = 0f,
    val blue: Float = 0f,
    val mode: Mode = Mode.COLOR
) {
    enum class Mode(val title: String) {
        COLOR("\uD83D\uDCA1 Color"),
        RAINBOW("\uD83C\uDF08 Rainbow")
    }

    val bytes: ByteArray
        get() {
            return byteArrayOf(
                red.toInt().toByte(),
                green.toInt().toByte(),
                blue.toInt().toByte(),
                if (mode == Mode.COLOR) 0.toByte() else 1.toByte()
            )
        }

    companion object {
        fun deserialize(data: ByteArray): LampState {
            var offset = 0
            val red = data[offset].toUByte()
            offset++
            val green = data[offset].toUByte()
            offset++
            val blue = data[offset].toUByte()
            offset++
            val mode = data[offset].toUByte()
            return LampState(
                red = red.toFloat(),
                green = green.toFloat(),
                blue = blue.toFloat(),
                mode = if (mode.toInt() == 0) Mode.COLOR else Mode.RAINBOW
            )
        }
    }
}
